File: CSharp-Project/DataStructure/Graph/Graph.cs
using Microsoft.VisualBasic;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection.Emit;
using System.Text;
using System.Threading.Tasks;

namespace Graph
{
    public class Graph
    {
        private class Vertice //Node
        {
            public String Lable { get; private set; }
            public Vertice(String lable) { Lable = lable; }
            public override String ToString() { return Lable; }
        }

        private Dictionary<String, Vertice> VerticeList { get; set; } //NodeList
        private Dictionary<Vertice, List<Vertice>> EdgeList { get; set; } //AdjecencyList
        public Graph() { VerticeList = new(); EdgeList = new(); }

        private bool IsNull(Vertice node) { return node == null; }

        public void AddNode(String label)
        {
            VerticeList.TryAdd(label, new Vertice(label));
            var node = VerticeList[label];
            EdgeList.TryAdd(node, new());  //new List
        }
        public void RemoveNode(String label)
        {
            var node = VerticeList[label];
            if (IsNull(node)) return;
            EdgeList.Remove(node);
            VerticeList.Remove(label);
        }

        public void AddEdge(String fromString, String toString)
        {   //relationship from -> to
            try { EdgeList[VerticeList[fromString]].Add(VerticeList[toString]); }
            catch (Exception) { throw new Exception(); }    //IsNull
        }
        public void RemoveEdge(String fromString, String toString)
        {   //remove from.to
            try { EdgeList[VerticeList[fromString]].Remove(VerticeList[toString]); }
            catch (Exception) { return; }    //IsNull
        }

        //Iteration
        public void TraverseDepthFirsy(String rootString)
        {
            HashSet<Vertice> visited = new();
            Stack<Vertice> stack = new();
            var root = VerticeList?[rootString];
            if (IsNull(root!)) return;
            stack.Push(root!);   //Link
            while (stack.Count != 0)
            {
                var node = stack.Pop(); //currernt
                if (visited.Contains(node)) continue;
                else visited.Add(node);
                Console.Write(node + " ");
                foreach (var neighbour in EdgeList[node])   //Links
                    if (!visited.Contains(neighbour)) stack.Push(neighbour);
            }
        }

        public void TraverseBreadthFirsy(String rootString)
        {
            HashSet<Vertice> visited = new();
            Queue<Vertice> queue = new();
            var root = VerticeList?[rootString];
            if (IsNull(root!)) return;
            queue.Enqueue(root!);   //Link
            while (queue.Count != 0)
            {
                var node = queue.Dequeue(); //currernt
                if (visited.Contains(node)) continue;
                else visited.Add(node);
                Console.Write(node + " ");
                foreach (var neighbour in EdgeList[node])   //Links
                    if (!visited.Contains(neighbour)) queue.Enqueue(neighbour);
            }
        }


        public void TraverseDepthFirsy_recursion(String rootString)
        {
            try 
            {
                Vertice root = VerticeList[rootString];
                TraverseDepthFirsy_recursion(root, new());
                Console.WriteLine();
            } catch (Exception) { return; }
        }
        private void TraverseDepthFirsy_recursion(Vertice node, HashSet<Vertice> visited)
        {
            Console.Write(node + " ");
            visited.Add(node);
            foreach (var neighbour in EdgeList[node])
                if (!visited.Contains(neighbour)) TraverseDepthFirsy_recursion(neighbour, visited);
        }




        public List<String> TopologicalSort()  //לחזור 
        {
            Stack<Vertice> stack = new();
            HashSet<Vertice> visitedSet = new();
            foreach (var node in VerticeList.Values) TopologicalSort(node, visitedSet, stack);
            return ToList(stack);
        }
        private List<String> ToList(Stack<Vertice> stack) {
            List<String> sortedList = new();
            while (stack.Count != 0) sortedList.Add(stack.Pop().Lable);
            return sortedList;
        }
        private void TopologicalSort(Vertice node, HashSet<Vertice> visitedSet, Stack<Vertice> stack)
        {
            if (visitedSet.Contains(node)) return;
            visitedSet.Add(node);
            foreach (var neighbour in EdgeList[node])
                TopologicalSort(neighbour, visitedSet, stack);
            stack.Push(node);
        }


        public bool HasCycle()
        {   //לחזור
            HashSet<Vertice> all = new();
            all.UnionWith(VerticeList.Values);// java: addAll //A.Concat(B).ToHashSet()
            HashSet<Vertice> visiting = new();
            HashSet<Vertice> visited = new();
            foreach(var current in all)  if (HasCycle(current, all, visiting, visited)) return true; ;
            
            return false;
        }
        private bool HasCycle(Vertice node, HashSet<Vertice> all, HashSet<Vertice> visiting, HashSet<Vertice> visited)
        {
            all.Remove(node);
            visiting.Add(node);
            foreach (var neighbour in EdgeList[node])
            {
                if (visited.Contains(neighbour)) continue;
                if (visiting.Contains(neighbour)) return true;
                if (HasCycle(neighbour, all, visiting, visited)) return true;
            }
            visiting.Remove(node);
            visited.Add(node);
            return false;
        }


        public override String ToString()
        {
            String str = "";
            foreach (var source in EdgeList.Keys)
            {
                var targets = EdgeList[source];
                if (targets.Count != 0) str += source + " is connected to [" + String.Join(", ", targets) + "]\n";
            }
            return str;
        }
    }
}

File: Cpp-Clion-Project/DataStructureNonLinear/Graph/Graph.cpp

#include <iostream>
#include <string>
#include <map>
#include <set>
#include <vector>
#include <list>
#include <stack>
#include <queue>
#include <memory>
#include <iterator>

using std::string, std::to_string, std::cout, std::map, std::pair, std::set;
using std::stack, std::queue, std::list, std::shared_ptr, std::make_shared, std::exception;

class Graph {   //for example only
private:
    struct Node {
     char label;
     explicit Node(char& label) : label{label} {   }
     ~Node(){cout << "del:" << label << " ";}
     bool operator<(const struct Node& other) const { return this->label < other.label; }
     char print() const { return label; }
    };

    typedef shared_ptr<struct Node> p_Node;
    typedef std::_Rb_tree_iterator<pair<const char, p_Node>> it_Node;

    map<char, p_Node> nodes;
    map<p_Node, list<p_Node>> edges;
    bool isNull(it_Node itNode) { return itNode == nodes.end();}
    void traverseDepthFirst_recursion(p_Node node, set<p_Node>& visited)
    {
        cout << node->print() << " ";
        visited.insert(node);
        for (auto neighbour : edges.find(node)->second)
        if (!visited.contains(neighbour)) traverseDepthFirst_recursion(neighbour, visited);
    }
    list<char> toList(stack<p_Node>& stack) {   //fix to insert with it
        list<char> sortedList{};
        while (stack.empty()) {sortedList.push_back(stack.top()->label); stack.pop();}
        return sortedList;
    }
    void topologicalSort(p_Node node, set<p_Node> visitedSet, stack<p_Node> stack)
    {
        if (visitedSet.contains(node)) return;
        visitedSet.insert(node);
        for (auto neighbour : edges.find(node)->second)
            topologicalSort(neighbour, visitedSet, stack);
        stack.push(node);
    }
    bool hasCycle(p_Node node, set<p_Node> all, set<p_Node> visiting, set<p_Node> visited)
    {
        all.erase(node);
        visiting.insert(node);
        for (auto neighbour : edges.find(node)->second)
        {
            if (visited.contains(neighbour)) continue;
            if (visiting.contains(neighbour)) return true;
            if (hasCycle(neighbour, all, visiting, visited)) return true;
        }
        visiting.erase(node);
        visited.insert(node);
        return false;
    }
public:
    explicit Graph() : nodes{}, edges{} {}
    ~Graph() {}
    void addNode(char label)
    {
        auto newNode= make_shared<Node>(label);
        nodes.insert({label, newNode});
        auto node = nodes.find(label)->second;
        list<p_Node> newNodeList{};
        edges.insert({node, newNodeList});  //new List
    }
    void removeNode(char label)
    {
        auto it = nodes.find(label);
        if ( it == nodes.end()) return;
        nodes.erase(label);
        edges.erase(it->second);
    }

    void addEdge(char from, char to)
    {   //relationship from -> to

        try
        {
            if(nodes.find(from) == nodes.end() || nodes.find(to) == nodes.end()) throw exception();
            edges.find(nodes.find(from)->second)->second.push_back(nodes.find(to)->second);
        } catch (...) { throw exception(); }    //IsNull
    }
    void removeEdge(char from, char to)
    {   //remove from.to
        try {edges.find(nodes.find(from)->second)->second.remove(nodes.find(to)->second);}
        catch (...) { return; }    //IsNull
    }



    //Iteration
    void traverseDepthFirst(char root)
    {
        set<p_Node> visited{};
        stack<p_Node> stack{};
        auto rootNode = nodes.find(root);
        if (isNull(rootNode)) return;
        stack.push(rootNode->second);   //Link
        while (stack.empty())
        {
            auto node = stack.top() ;
            stack.pop();
            if (visited.contains(node)) continue;
            else visited.insert(node);
            cout << node << " ";
            for (auto neighbour : edges.find(node)->second)   //Links
                if (!visited.contains(neighbour)) stack.push(neighbour);
        }
    }

    void traverseBreadthFirst(char root)
    {
        set<p_Node> visited{};
        queue<p_Node> queue{};
        auto rootNode = nodes.find(root);
        if (isNull(rootNode)) return;
        queue.push(rootNode->second);   //Link
        while (queue.empty())
        {
            auto node = queue.front() ;
            queue.pop();
            if (visited.contains(node)) continue;
            else visited.insert(node);
            cout << node << " ";
            for (auto neighbour : edges.find(node)->second)   //Links
                if (!visited.contains(neighbour)) queue.push(neighbour);
        }
    }

    void traverseDepthFirst_recursion(char root)
    {
        try
        {
            set<p_Node> visited{};
            auto rootNode = nodes.find(root)->second;
            traverseDepthFirst_recursion(rootNode, visited);
            cout << "\n";
        } catch (...) { return; }
    }

    bool hasCycle()
    {   //לחזור
        set<p_Node> allNodes{};
                for(auto it : nodes) { allNodes.insert(it.second);}
        set<p_Node> visiting{};
        set<p_Node> visited{};
        for(auto current : allNodes)  if (hasCycle(current, allNodes, visiting, visited)) return true;
        return false;
    }


    string print()
    {
        string str = "";
        for (auto itMap : edges)
        {
            auto targets = edges.find(itMap.first)->second;
            if (!targets.empty()) {
                str += itMap.first->print();
                str += " is connected to [";
                for(auto itList: targets) {
                    if(itList == NULL) continue;
                    str += itList->print();
                    str +=  ", ";
                }
                str += "]\n";
            }
        }
        return str;
    }

};